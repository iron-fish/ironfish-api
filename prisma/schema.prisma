generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["interactiveTransactions"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Event {
  id          Int       @id @default(autoincrement())
  created_at  DateTime  @default(now()) @db.Timestamp(6)
  updated_at  DateTime  @default(now()) @updatedAt @db.Timestamp(6)
  type        EventType
  occurred_at DateTime  @db.Timestamp(6)
  points      Int
  user_id     Int
  deleted_at  DateTime? @db.Timestamp(6)
  block_id    Int?      @unique(map: "uq_events_on_block_id")
  block       Block?    @relation(fields: [block_id], references: [id])
  user        User      @relation(fields: [user_id], references: [id])
  url         String?   @db.VarChar

  @@index([block_id], name: "index_events_on_block_id")
  @@index([user_id], name: "index_events_on_user_id")
  @@map("events")
}

model User {
  id                  Int         @id @default(autoincrement())
  created_at          DateTime    @default(now()) @db.Timestamp(6)
  updated_at          DateTime    @default(now()) @updatedAt @db.Timestamp(6)
  email               String      @unique(map: "uq_users_on_email") @db.VarChar
  graffiti            String      @unique(map: "uq_users_on_graffiti") @db.VarChar
  total_points        Int         @default(0)
  country_code        String      @db.VarChar
  email_notifications Boolean     @default(false)
  last_login_at       DateTime?   @db.Timestamp(6)
  discord             String?     @db.VarChar
  telegram            String?     @db.VarChar
  github              String?     @db.VarChar
  events              Event[]
  points              UserPoints?
  node_uptime         NodeUptime?

  @@index([email], name: "index_users_on_email")
  @@index([graffiti], name: "index_users_on_graffiti")
  @@map("users")
}

model NodeUptime {
  id              Int      @id @default(autoincrement())
  user            User     @relation(fields: [user_id], references: [id])
  user_id         Int
  total_hours     Int      @default(0)
  last_checked_in DateTime @default(now()) @db.Timestamp(6)

  @@index([user_id], name: "index_node_uptime_on_user_id")
  @@map("node_uptimes")
}

model Block {
  id                       Int                @id @default(autoincrement())
  created_at               DateTime           @default(now()) @db.Timestamp(6)
  updated_at               DateTime           @default(now()) @updatedAt @db.Timestamp(6)
  hash                     String             @db.VarChar
  sequence                 Int
  previous_block_hash      String?            @db.VarChar
  main                     Boolean
  network_version          Int
  transactions_count       Int
  timestamp                DateTime           @db.Timestamp(6)
  time_since_last_block_ms Int?
  graffiti                 String             @db.VarChar
  size                     Int?
  difficulty               BigInt
  blocks_transactions      BlockTransaction[]
  event                    Event?

  @@unique([hash, network_version], name: "uq_blocks_on_hash_and_network_version", map: "uq_blocks_on_hash_and_network_version")
  @@index([hash], name: "index_blocks_on_hash")
  @@index([hash, network_version], name: "index_blocks_on_hash_and_network_version")
  @@index([sequence, network_version], name: "index_blocks_on_sequence_and_network_version")
  @@map("blocks")
}

model Transaction {
  id                  Int                @id @default(autoincrement())
  created_at          DateTime           @default(now()) @db.Timestamp(6)
  updated_at          DateTime           @default(now()) @updatedAt @db.Timestamp(6)
  hash                String             @db.VarChar
  network_version     Int
  fee                 Int
  size                Int
  notes               Json
  spends              Json
  blocks_transactions BlockTransaction[]

  @@unique([hash, network_version], name: "uq_transactions_on_hash_and_network_version", map: "uq_transactions_on_hash_and_network_version")
  @@index([hash], name: "index_transactions_on_hash")
  @@index([hash, network_version], name: "index_transactions_on_hash_and_network_version")
  @@map("transactions")
}

model BlockTransaction {
  block          Block       @relation(fields: [block_id], references: [id])
  block_id       Int
  transaction    Transaction @relation(fields: [transaction_id], references: [id])
  transaction_id Int

  @@id([block_id, transaction_id])
  @@index([block_id], name: "index_blocks_transactions_on_block_id")
  @@index([transaction_id], name: "index_blocks_transactions_on_transaction_id")
  @@map("blocks_transactions")
}

model FaucetTransaction {
  id           Int       @id @default(autoincrement())
  created_at   DateTime  @default(now()) @db.Timestamp(6)
  updated_at   DateTime  @default(now()) @updatedAt @db.Timestamp(6)
  email        String?   @db.VarChar
  public_key   String    @db.VarChar
  started_at   DateTime? @db.Timestamp(6)
  completed_at DateTime? @db.Timestamp(6)
  tries        Int       @default(0)
  hash         String?   @db.VarChar

  @@index([email], map: "index_faucet_transactions_on_email")
  @@index([public_key], map: "index_faucet_transactions_on_public_key")
  @@index([hash], map: "index_faucet_transactions_on_hash")
  @@map("faucet_transactions")
}

model BlockDaily {
  id                         Int      @id @default(autoincrement())
  created_at                 DateTime @default(now()) @db.Timestamp(6)
  updated_at                 DateTime @default(now()) @updatedAt @db.Timestamp(6)
  date                       DateTime @db.Timestamp(6) @unique(map: "uq_blocks_daily_on_date")
  unique_graffiti_count      Int
  average_block_time_ms      Int
  blocks_count               Int
  blocks_with_graffiti_count Int
  cumulative_unique_graffiti Int
  transactions_count         Int
  average_difficulty_millis  BigInt
  chain_sequence             Int

  @@index([date], map: "index_blocks_daily_on_date")
  @@map("blocks_daily")
}

model Version {
  id         Int      @id @default(autoincrement())
  created_at DateTime @default(now()) @db.Timestamp(6)
  version    String   @db.VarChar

  @@map("versions")
}

enum EventType {
  BLOCK_MINED
  BUG_CAUGHT
  COMMUNITY_CONTRIBUTION
  PULL_REQUEST_MERGED
  SOCIAL_MEDIA_PROMOTION

  @@map("event_type")
}

model UserPoints {
  id                                             Int       @id @default(autoincrement())
  created_at                                     DateTime  @default(now()) @db.Timestamp(6)
  updated_at                                     DateTime  @default(now()) @updatedAt @db.Timestamp(6)
  user_id                                        Int       @unique
  total_points                                   Int       @default(0)
  block_mined_points                             Int       @default(0)
  block_mined_last_occurred_at                   DateTime? @db.Timestamp(6)
  bug_caught_points                              Int       @default(0)
  bug_caught_last_occurred_at                    DateTime? @db.Timestamp(6)
  community_contribution_points                  Int       @default(0)
  community_contribution_last_occurred_at        DateTime? @db.Timestamp(6)
  pull_request_merged_points                     Int       @default(0)
  pull_request_merged_last_occurred_at           DateTime? @db.Timestamp(6)
  social_media_promotion_points                  Int       @default(0)
  social_media_promotion_last_occurred_at        DateTime? @db.Timestamp(6)
  user                                           User      @relation(fields: [user_id], references: [id])

  @@index([user_id], map: "index_user_points_on_user_id")
  @@index([block_mined_points], map: "index_user_points_on_block_mined")
  @@index([bug_caught_points], map: "index_user_points_on_bug_caught")
  @@index([community_contribution_points], map: "index_user_points_on_community_contribution")
  @@index([pull_request_merged_points], map: "index_user_points_on_pull_request_merged")
  @@index([social_media_promotion_points], map: "index_user_points_on_social_media_promotion")
  @@map("user_points")
}
