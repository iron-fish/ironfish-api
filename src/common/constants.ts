/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { EventType } from '.prisma/client';

export const DAYS_IN_WEEK = 7;
export const MS_PER_DAY = 1000 * 60 * 60 * 24;

// Pagination limits
export const DEFAULT_LIMIT = 20;
export const MAX_LIMIT = 100;

// Event categories
export const WEEKLY_POINT_LIMITS_BY_EVENT_TYPE: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 1000,
  [EventType.BUG_CAUGHT]: 1000,
  [EventType.COMMUNITY_CONTRIBUTION]: 2000,
  [EventType.PULL_REQUEST_MERGED]: 5000,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 1000,
};

export const POINTS_PER_CATEGORY: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 100,
  [EventType.BUG_CAUGHT]: 100,
  [EventType.COMMUNITY_CONTRIBUTION]: 1000,
  [EventType.PULL_REQUEST_MERGED]: 500,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 100,
};
