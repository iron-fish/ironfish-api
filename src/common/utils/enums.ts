/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

export type StrEnumValue<T> = T[keyof T];
export type StrEnum<T> = Record<keyof T, string>;

function getEnumStrValues<T extends StrEnum<T>>(
  enumType: T,
): Array<StrEnumValue<T>> {
  return Object.values(enumType)
    .filter((v) => typeof v === 'string')
    .map((v) => v as StrEnumValue<T>);
}

export function isInStrEnum<T extends StrEnum<T>>(
  value: unknown,
  enumType: T,
): boolean {
  return !!getEnumStrValues(enumType).find((enumValue) => enumValue === value);
}
