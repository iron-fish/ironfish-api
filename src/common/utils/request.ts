/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Request } from 'express';

export function fetchIpAddressFromRequest(request: Request): string {
  const xForwardedFor = request.header('X-Forwarded-For');
  if (xForwardedFor) {
    const addresses = xForwardedFor.split(',');
    if (addresses.length) {
      return addresses[0].trim();
    }
  }

  // In Heroku, `X-Forwarded-For` will always be set. However, this will be used
  // as a fallback in local development.
  return request.ip;
}
