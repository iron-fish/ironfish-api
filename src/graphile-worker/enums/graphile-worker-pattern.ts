/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
export enum GraphileWorkerPattern {
  DELETE_BLOCK_MINED_EVENT = 'DELETE_BLOCK_MINED_EVENT',
  CREATE_NODE_UPTIME_EVENT = 'CREATE_NODE_UPTIME_EVENT',
  REFRESH_USERS_POINTS = 'REFRESH_USERS_POINTS',
  REFRESH_USER_POINTS = 'REFRESH_USER_POINTS',
  SYNC_BLOCKS_DAILY = 'SYNC_BLOCKS_DAILY',
  UPDATE_LATEST_POINTS = 'UPDATE_LATEST_POINTS',
  UPSERT_BLOCK_MINED_EVENT = 'UPSERT_BLOCK_MINED_EVENT',
}
