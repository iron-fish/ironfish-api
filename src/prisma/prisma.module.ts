/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { ApiConfigModule } from '../api-config/api-config.module';
import { PrismaService } from './prisma.service';

@Module({
  exports: [PrismaService],
  imports: [ApiConfigModule],
  providers: [PrismaService],
})
export class PrismaModule {}
