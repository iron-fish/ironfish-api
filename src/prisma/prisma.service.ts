/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';
import { ApiConfigService } from '../api-config/api-config.service';

@Injectable()
export class PrismaService
  extends PrismaClient
  implements OnModuleInit, OnModuleDestroy
{
  constructor(readonly config: ApiConfigService) {
    super({
      datasources: {
        db: {
          url: config.dbPoolUrl,
        },
      },
      // uncomment to log queries
      // log: ['query', 'info', 'warn', 'error'],
    });
  }

  async onModuleInit(): Promise<void> {
    await this.$connect();
  }

  async onModuleDestroy(): Promise<void> {
    await this.$disconnect();
  }
}
