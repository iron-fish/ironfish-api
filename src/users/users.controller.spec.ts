/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common';
import faker from 'faker';
import request from 'supertest';
import { ulid } from 'ulid';
import { v4 as uuid } from 'uuid';
import { PrismaService } from '../prisma/prisma.service';
import { bootstrapTestApp } from '../test/test-app';
import { MetricsGranularity } from './enums/metrics-granularity';

const API_KEY = 'test';

describe('UsersController', () => {
  let app: INestApplication;
  let prisma: PrismaService;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    prisma = app.get(PrismaService);
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('GET /users/:id', () => {
    describe('with a valid id', () => {
      it('returns the user', async () => {
        const user = await prisma.user.create({
          data: {
            confirmation_token: ulid(),
            confirmed_at: new Date().toISOString(),
            email: faker.internet.email(),
            graffiti: uuid(),
            country_code: faker.address.countryCode(),
          },
        });
        const { body } = await request(app.getHttpServer())
          .get(`/users/${user.id}`)
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          id: user.id,
          graffiti: user.graffiti,
          total_points: expect.any(Number),
        });
      });
    });

    describe('with a missing id', () => {
      it('returns a 404', async () => {
        await request(app.getHttpServer())
          .get('/users/12345')
          .expect(HttpStatus.NOT_FOUND);
      });
    });
  });

  describe('GET /users/:id/metrics', () => {
    describe('with start but no end', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/users/123/metrics')
          .query({
            start: new Date().toISOString(),
            granularity: MetricsGranularity.TOTAL,
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with end but no start', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/users/123/metrics')
          .query({
            end: new Date().toISOString(),
            granularity: MetricsGranularity.TOTAL,
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a missing granularity', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/users/123/metrics')
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a time range for a LIFETIME request', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/users/123/metrics')
          .query({
            start: new Date().toISOString(),
            end: new Date().toISOString(),
            granularity: MetricsGranularity.LIFETIME,
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with start after end', () => {
      it('returns a 422', async () => {
        const start = new Date().toISOString();
        const end = new Date(Date.now() - 1).toISOString();
        const { body } = await request(app.getHttpServer())
          .get('/users/123/metrics')
          .query({
            start,
            end,
            granularity: MetricsGranularity.TOTAL,
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a time range longer than the supported range', () => {
      it('returns a 422', async () => {
        const start = '2021-06-01T00:00:00.000Z';
        const end = '2021-08-01T00:00:00.000Z';
        const { body } = await request(app.getHttpServer())
          .get('/users/123/metrics')
          .query({
            start,
            end,
            granularity: MetricsGranularity.TOTAL,
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a TOTAL request and no time range', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/users/123/metrics')
          .query({
            granularity: MetricsGranularity.TOTAL,
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a missing user', () => {
      it('returns a 404', async () => {
        const start = new Date(Date.now() - 1).toISOString();
        const end = new Date().toISOString();
        await request(app.getHttpServer())
          .get('/users/12345/metrics')
          .query({
            start,
            end,
            granularity: MetricsGranularity.TOTAL,
          })
          .expect(HttpStatus.NOT_FOUND);
      });
    });

    describe('with a valid lifetime request', () => {
      it('returns the lifetime metrics for the user', async () => {
        const user = await prisma.user.create({
          data: {
            confirmation_token: ulid(),
            confirmed_at: new Date().toISOString(),
            email: faker.internet.email(),
            graffiti: uuid(),
            country_code: faker.address.countryCode('alpha-3'),
          },
        });
        const granularity = MetricsGranularity.LIFETIME;

        const { body } = await request(app.getHttpServer())
          .get(`/users/${user.id}/metrics`)
          .query({
            granularity,
          });
        expect(body).toMatchObject({
          user_id: user.id,
          granularity,
          points: expect.any(Number),
          metrics: {
            blocks_mined: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
            bugs_caught: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
            community_contributions: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
            pull_requests_merged: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
            social_media_contributions: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
          },
        });
      });
    });

    describe('with a valid total request', () => {
      it('returns the total metrics for the user in the given range', async () => {
        const user = await prisma.user.create({
          data: {
            confirmation_token: ulid(),
            confirmed_at: new Date().toISOString(),
            email: faker.internet.email(),
            graffiti: uuid(),
            country_code: faker.address.countryCode('alpha-3'),
          },
        });
        const start = new Date(Date.now() - 1).toISOString();
        const end = new Date().toISOString();
        const granularity = MetricsGranularity.TOTAL;

        const { body } = await request(app.getHttpServer())
          .get(`/users/${user.id}/metrics`)
          .query({
            granularity,
            start,
            end,
          });
        expect(body).toMatchObject({
          user_id: user.id,
          granularity,
          points: expect.any(Number),
          metrics: {
            blocks_mined: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
            bugs_caught: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
            community_contributions: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
            pull_requests_merged: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
            social_media_contributions: {
              count: expect.any(Number),
              points: expect.any(Number),
            },
          },
        });
      });
    });
  });

  describe('GET /users', () => {
    describe('with an invalid order by option', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/users')
          .query({
            order_by: 'foobar',
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a valid request', () => {
      it('returns a list of users', async () => {
        const { body } = await request(app.getHttpServer())
          .get(`/users`)
          .expect(HttpStatus.OK);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(0);
        expect((data as unknown[])[0]).toMatchObject({
          id: expect.any(Number),
          graffiti: expect.any(String),
        });
      });
    });

    describe('with `order_by` provided', () => {
      it('returns ranks with the users (and no country_code provided)', async () => {
        const { body } = await request(app.getHttpServer())
          .get(`/users`)
          .query({ order_by: 'rank' })
          .expect(HttpStatus.OK);

        const { data } = body;
        expect((data as unknown[]).length).toBeGreaterThan(0);
        expect((data as unknown[])[0]).toMatchObject({
          id: expect.any(Number),
          graffiti: expect.any(String),
          rank: expect.any(Number),
        });
      });
    });
    const getLocation = async () => {
      const { body } = await request(app.getHttpServer())
        .get('/users')
        .expect(HttpStatus.OK);

      const { data } = body;
      const places = (data as Record<string, unknown>[]).map(
        ({ country_code: cc }) => cc,
      );
      return places[0] as string;
    };
    describe('with `country_code` provided', () => {
      it('allows filtering by country_code', async () => {
        const someplace = await getLocation();
        const { body } = await request(app.getHttpServer())
          .get('/users')
          .query({ country_code: someplace })
          .expect(HttpStatus.OK);
        const { data } = body;
        (data as Record<string, unknown>[]).map(({ country_code }) =>
          expect(country_code).toEqual(someplace),
        );
      });
    });

    describe('when rank is provided', () => {
      it('filters by country code', async () => {
        const someplace = await getLocation();
        const { body } = await request(app.getHttpServer())
          .get('/users')
          .query({ country_code: someplace, order_by: 'rank' })
          .expect(HttpStatus.OK);
        const { data } = body;
        (data as Record<string, unknown>[]).map(({ country_code }) =>
          expect(country_code).toEqual(someplace),
        );
      });
    });

    describe('with an invalid cursor', () => {
      it('returns a 404', async () => {
        await request(app.getHttpServer())
          .get(`/users`)
          .query({ order_by: 'rank', before: -1 })
          .expect(HttpStatus.NOT_FOUND);
      });
    });
  });

  describe('POST /users', () => {
    describe('with missing arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post(`/users`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with empty arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post(`/users`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ email: '', graffiti: '' })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a duplicate email', () => {
      it('returns a 422', async () => {
        const user = await prisma.user.create({
          data: {
            confirmation_token: ulid(),
            email: faker.internet.email(),
            graffiti: uuid(),
            country_code: faker.address.countryCode('alpha-3'),
          },
        });
        await request(app.getHttpServer())
          .post(`/users`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ email: user.email, graffiti: user.graffiti })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);
      });
    });

    describe('with valid arguments', () => {
      it('creates an user', async () => {
        const email = faker.internet.email();
        const graffiti = uuid();
        const discord = faker.internet.userName();
        const { body } = await request(app.getHttpServer())
          .post(`/users`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({
            email,
            graffiti,
            discord,
            country_code: faker.address.countryCode('alpha-3'),
          })
          .expect(HttpStatus.CREATED);
        expect(body).toMatchObject({
          id: expect.any(Number),
          email,
          graffiti,
          discord,
        });
      });
    });
  });
});
