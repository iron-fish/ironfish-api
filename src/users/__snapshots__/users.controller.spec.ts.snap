// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersController GET /users with an invalid order by option returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "order_by must be equal to total_points",
  ],
  "statusCode": 422,
}
`;

exports[`UsersController GET /users/:id/metrics with a TOTAL request and no time range returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "Must provide time range for \\"TOTAL\\" requests",
  "statusCode": 422,
}
`;

exports[`UsersController GET /users/:id/metrics with a missing granularity returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "granularity must be a valid enum value",
  ],
  "statusCode": 422,
}
`;

exports[`UsersController GET /users/:id/metrics with a time range for a LIFETIME request returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "Cannot provide time range for \\"LIFETIME\\" requests",
  "statusCode": 422,
}
`;

exports[`UsersController GET /users/:id/metrics with a time range longer than the supported range returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "Time range too long",
  "statusCode": 422,
}
`;

exports[`UsersController GET /users/:id/metrics with end but no start returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "start must be a Date instance",
  ],
  "statusCode": 422,
}
`;

exports[`UsersController GET /users/:id/metrics with start after end returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "\\"start\\" must be stricly less than \\"end\\"",
  "statusCode": 422,
}
`;

exports[`UsersController GET /users/:id/metrics with start but no end returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "end must be a Date instance",
  ],
  "statusCode": 422,
}
`;

exports[`UsersController POST /users with missing arguments returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "country_code must be a valid ISO31661 Alpha3 code",
    "email must be a string",
    "graffiti must be a string",
  ],
  "statusCode": 422,
}
`;
